Add option to view feed without loading feeds from a config

Nemoden opened the pull request:
Hey there!

This deals with issue #13 

The proposed solution is to add a `--preview` flag, but I find it unnecessary as

```
nom <URL>
```

feels like a more straightforward interface since we can check if `args[0]` is a URL or not and as such distinguish vs commands

idk if it's a good solution, I have 0 experience with go and 0 experience contributing to open source projects. But I was looking forward to make my first one not only to learn go, but contribute to something that I actually looking to use (long time newsboat user) and a feature to preview a feed looked like a nice opportunity as it was something I was missing in newsboat.

The solution doesn't use cache, and as such a little slow on transitions, - maybe a short-lived in-memory cache can be implemented here. 

Code review is very much appreciated!

Cheers.

Demo



https://user-images.githubusercontent.com/795034/213139202-b3aa7be3-890e-414c-896e-121bed46efdd.mov

guyfedwards commented:
Hey, thanks for the contribution. I agree it would be simpler in the case of a single URL, but it also doesn't cover all usecases and breaks the pattern of `nom <subcommand> [config flags]` which I find more intuitive.

If you are interested in refactoring this to use a `--flags` argument instead that would override the config.flags value if passed I would love to get that merged. This way it will utilise the cache and all other mechanisms under the hood without any additional work.

End goal would be: 
```
$ nom --feed=foo.rss.com
$ nom --feed=bar.rss.com --feed=baz.rss.com
```

Let me know if taht makes sense and/or if you are interested in making the change.

guyfedwards commented on internal/commands/main.go:
```diff
@@ -90,11 +90,18 @@ type FetchResultError struct {
 		wg   sync.WaitGroup
 	)
 
+	if c.config.Preview != "" {
```
If we change this to just set config.feeds based on the preview url, then we can use the existing fetching/caching logic.

```go
   feeds := c.config.Feeds

   // if we have a preview url, ignore the config URLs
   if c.config.Preview != nil {
     feeds := []Feed{URL: c.config.Preview}
   }
```

guyfedwards commented:
I've been thinking about this a bit more and whilst I still want to add the `--feed` option for config, I don't see why this can't also be present as it is a very convenient way to view a single rss feed quickly.

guyfedwards commented on internal/commands/main.go:
```diff
@@ -162,22 +169,14 @@ func (c Commands) FindArticle(substr string) (item rss.Item, err error) {
-				item = it
-				break
-			}
+		item, err := r.FindArticle(substr)
```
What was the reasoning for moving this logic?

Nemoden replied:
> I've been thinking about this a bit more and whilst I still want to add the `--feed` option for config, I don't see why this can't also be present as it is a very convenient way to view a single rss feed quickly.

Yep, even something like

```
nom url1 url2 url3
```

as a shortcut for

```
nom --feed=url1 --feed=url2 --feed=url3
```

Anyway, it's the just the interface, how the info gets in, udnrelying mechanics of how feeds are previewed should be the same.

Thank you for the code review! Very helpful. Let's see what I can do

Nemoden replied:
@guyfedwards initial implementation was a little different to what you see on the PR and it required being able to find an article in a single instance of `RSS`.

However, later on, I've changed the business logic so it'd "hook up" better into existing code yet decided to leave this helper method as generally useful. Happy to revert this change tho, - without the context of what my initial implementation was it indeed doesn't make sense :)

guyfedwards replied:
Makes sense. I would revert it for this PR to keep the changes isolated.

guyfedwards commented on internal/commands/main.go:
```diff
@@ -162,22 +169,14 @@ func (c Commands) FindArticle(substr string) (item rss.Item, err error) {
-			}
+		item, err := r.FindArticle(substr)
+		if item != nil || err != nil {
+			return *item, err
```
General pattern in go is to check for an error and explitly return, `<zero_value>, err` if therre is an error.

This would be better more explicitly as:
```go
if err != nil {
  return rss.Item{}, err 
}

if item == nil {
  return rss.Item{}, errors.New(fmt.Sprintf("Article %s not found", substr))
}
...
return *item, nil
```

Nemoden replied:
@guyfedwards noted. Cheers!

Nemoden commented:
> I've been thinking about this a bit more and whilst I still want to add the `--feed` option for config, I don't see why this can't also be present as it is a very convenient way to view a single rss feed quickly.

Yep, even something like

```
nom url1 url2 url3
```

as a shortcut for

```
nom --feed=url1 --feed=url2 --feed=url3
```

Anyway, it's the just the interface, how the info gets in, udnrelying mechanics of how feeds are previewed should not be dependant on the interface, then any kind of input interface can be wired up.

Thank you for the code review! Very helpful. Let's see what I can do

Nemoden commented:
@guyfedwards 

Updated the code, and added a little demo.

Basically preview URLs can be specified BOTH via the --feed flag OR without the flag, - any URL on the CLI is considered a feed to preview. And both can work at the same time, i.e. this

```
nom --feed=https://example.com/rss.xml --feed=https://acme.com/atom.xml https://test.corp.dev/rss.xml
```

is legit, it's 3 URLs that will be previewed

I still haven't implemented short-lived cache, so it feels sluggish, but it can be tackled later.

The main reason why I haven't reused `Config.Feeds` is because inadvertent `config.Write` while in preview context will replace all the user's feeds in their main config file, but maybe I'm overlooking something? Potentially can be covered with a test, but I was looking to play it safe and semantically `config.Feeds` are the feeds a user has added to their config, whilst the preview feeds are the feeds that should always be discarded on the program exit. Felt wrong to conflate those, save for possible config override.

Nemoden commented:
Sorry, this PR is breaking `nom <command> <URL>`, will come up with the remediation soon

guyfedwards commented on cmd/nom/main.go:
```diff
@@ -13,17 +13,27 @@ import (
+		// Preview mode should use short-lived in-momory cache, but to make it
+		// happen we should support that through a cache interface.
+		// Another solution is to cache in different directory and wipe it out once nom is closed, i.e.
+		// defer wipeOutPreviewCache()
```
Not sure we really need to cache preview feeds. By design they are previews and thus should pull the live, if users want to have content cached, the feeds should be permanent.

If you agree, we can remove this comment

Nemoden replied:
@guyfedwards Ahmmm, the problem with that is the fact that we do `fetchAllFeeds` on any `FindArticle` which does another HTTP request per feed, which is completely unnecessary since we are getting the same information we already fetched in the current session.

Let's say we are previewing 2 feeds:

- acme.com/rss.xml
- example.com/rss.xml

On the startup, we do 2 HTTP calls to get those feeds' contents.

Once user navigates to one of the articles, we do `FindArticle` which calls `fetchAllFeeds` resulting in 2 unnecessary HTTP calls. The more the user views articles, the more calls we are making in vain.

The repercussions are not only that the interface is sluggish, but we are risking requesting some of the feeds too often and getting rate-limited.

I've been playing around with that and here is what I came up with https://github.com/Nemoden/nom/commit/5c9c026e3199acd6f6679c205ffcd92d62f9d29a

But, I'm not sure how the design was envisioned and whether using caching mechanism is conceptually correct here.

Please let me know your thought around this.

guyfedwards replied:
Ah I think I see the issue. The caching implementation was very basic and hidden behind fetch. Lets leave this comment in for now and can address teh caching for previews issue separately. Created https://github.com/guyfedwards/nom/issues/30 to track this

guyfedwards commented on internal/config/main.go:
```diff
@@ -97,6 +109,13 @@ func (c *Config) AddFeed(feed Feed) error {
 	return nil
 }
 
+func (c *Config) GetFeeds() []Feed {
```
Convention in Go is to have getters as public and values as private. So in this instance we would have:

```go
type Config struct {
  feeds []Feed
  Feeds func() []Feed
}
```

(Will need to update anywhere that references `config.Feeds` to `config.Feeds()`)

See here for more explanation https://go.dev/doc/effective_go#Getters

Nemoden replied:
@guyfedwards Thanks! Makes sense. So if that's the case, should I then write up a custom marshaller/unmarshaller for the config to marshal/unmarshal now unexported `feeds` field?

guyfedwards replied:
Bah, I forgot about that. Whilst I think that would be the more "Go" way to do it, it feels like an unneccessary complexity to add to the codebase vs a clear "GetFeeds" so am happy to leave as is

guyfedwards reviewed (commented):
Looks great. I can squash on merge so no need to worry. Let me know when these comments addressed and I'll push the button :+1:

Nemoden commented:
> Looks like tests are failing as there is now a new parameter for New, fix them and I can merge @Nemoden

👍 Just did that, added new tests, pushed, didn't squash myself

guyfedwards commented:
Thanks @Nemoden for your help and work on this!

guyfedwards commented:
Released in https://github.com/guyfedwards/nom/releases/tag/v1.1.6 🎉

Nemoden commented:
> Thanks @Nemoden for your help and work on this!

🍻
